Post on IDS Tutorials 2020/21 (#19)
mine-cetinkaya-rundel/citizenstatistician: content/post/2020-07-24-data-science-tutorials-with-learnr-and-gradethis.md

---

> title: Data Science Tutorials with Learnr and Gradethis

Reviewer: ```suggestion
title: Data Science Tutorials with learnr and gradethis
```

author: Lee Suddaby, Zeno Kujawa
date: '2020-07-24'
slug: data-science-tutorials-with-learnr-and-gradethis
categories: 
  - teaching
  - shiny
  - rstats

> tags: []

Reviewer: ```suggestion
tags: 
  - learnr
  - gradethis
  - tidyverse
```

keywords:
  - tech
---

> *This post was contributed by [Lee Suddaby](https://github.com/lee-suddaby) and [Zeno Kujawa](https://github.com/ZenoMK).*

Reviewer: Feel free to add one more sentence about each of you here if you like, like you're students at UoE, year, major, etc.

> Over the university summer break, we (Zeno and Lee) were busy making preparations for moving more of our Introduction to Data Science course from being human-graded to computer-graded. 

Reviewer: ```suggestion
Over the university summer break, we (Zeno and Lee) were busy making preparations for moving more of our [Introduction to Data Science](https://introds.org/) course from being human-graded to computer-graded. 
```

Reviewer: Maybe right here you can also add something about the fact that you took this course in the Fall and learned R there.

<!--more-->

Moving course delivery online and to more automated marking has definite advantages. For one, feedback can be instant and significantly more informative than R's sometimes cryptic errors: for example, what on earth does `"object of type 'closure' is not subsettable"` mean? For two, this method is much more scalable. Courses are growing, which *should* be a good thing, but manual marking is not scalable unless your human resources are growing at the same rate as class sizes.

> The solution then? Interactive tutorials, with help from the `shiny`, `learnr` and `gradethis`. In our case, we were converting written homeworks (the very ones we'd both completed on our time on the course last fall!) into the auto-graded tutorials.

Reviewer: ```suggestion
The solution then? Interactive tutorials, with help from the [**learnr**](https://rstudio.github.io/learnr/) and [**gradethis**](https://github.com/rstudio-education/gradethis) packages. In our case, we were converting written homeworks (the very ones we'd both completed on our time on the course last fall!) into the auto-graded tutorials.
```

### Work(flow)ing 9 to 5

> Behind every successful project is a successful workflow, so that is where we start. Our base is, of course, Github. Then, whisk in a handful of Pull Requests, two cups of peer reviews, and a few Issues if necessary, and mix to form a smooth batter. Bake in an RStudio environment for 20-25 minutes and add Merge Conflicts to taste.

Reviewer: ```suggestion
Behind every successful project is a successful workflow, so that is where we start. Our base is, of course, Github. Then, whisk in a handful of pull requests, two cups of peer reviews, and a few issues if necessary, and mix to form a smooth batter. Bake in an RStudio environment for 20-25 minutes and add merge conflicts to taste.
```

> Our first task was to familiarise ourselves with the particular packages - `learnr` and `gradethis` (more on these two below). Having done this, our workflow involved writing the tutorials two at a time - one person writing each - then reviewing the other person's work. After as many cycles of implementing and reviewing changes, a final review was carried out by the course organiser Dr. Çetinkaya-Rundel, who is also the creator of Citizen Statistician.

Reviewer: No need to use monospace text for package names going forward. I bolded them in their first occurrence to draw attention.

### learnr

`learnr` is an R package designed to make it easy to build interactive tutorials in R markdown documents. These tutorials can contain all manner of objects, including text, graphics and figures, sections and subsections, code exercises that students can execute, multiple choice questions, even YouTube videos!

> If you're interested in using (or just playing about with) the package yourself, it's available on CRAN.

Reviewer: add something to the end like "and you can learn more about it here" with "here" linking to the package website.

### gradethis

`gradethis` is a companion to the `learnr` package, and provides multiple different functions for grading learnr exercises. Code can either be graded against a certain solution, where we check that the student's answer exactly matches the solution, or the result can be checked against a set of conditions.
With the gradethis package, it's also easy to give helpful feedback if a student gets a foreseen wrong answer, so we can nudge them in the right direction.

This package is not yet on CRAN, but you can install it from the Github repository [here](https://github.com/rstudio-education/gradethis).

### Challenges

> We quickly learned that converting the homework assignments wouldn't be as straightforward as just replicating the questions in a `shiny` document. Since some tasks like interpreting visualisations can't easily be graded automatically, we had to make certain trade-offs. Usually, this either meant replacing an open-ended question with a series of multiple-choice questions while trying to preserve a questions essential lesson. 

Reviewer: ```suggestion
We quickly learned that converting the homework assignments wouldn't be as straightforward as just replicating the questions in a learnr tutorial. Since some tasks like interpreting visualisations can't easily be graded automatically, we had to make certain trade-offs. Usually, this either meant replacing an open-ended question with a series of multiple-choice questions while trying to preserve a questions essential lesson. 
```

> Another question we had to ask ourselves frequently was whether the students code itself or the output it produced should be evaluated. Here's an example of this conundrum: 

Reviewer: ```suggestion
Another question we had to ask ourselves frequently was whether the student's code itself or the output it produced should be evaluated. Here's an example of this conundrum: 
```

> <img src="https://imgur.com/sVa8ehy.jpg" width="499" height="95" />

Reviewer: These images are not showing up on the rendered document for some reason. I can't tell why, but if you don't manage to get it to work I'm happy to take another look at it and fix it before the post goes live.

Author: I've fixed that. The images are now in the `static` folder of the repo instead of hosted on Imgur.

With visualisations and code that doesn't produce any output, however, this becomes a bit trickier. 
In those situations, we decided to grade the code itself, which has the disadvantage that the code has to be exactly the same as in the solution, even though it would have the same effect if two lines were interchanged. We tried to circumvent the problem by giving the students more hints and providing a structure where fewer mistakes would be possible. The following image is an example of how we dealt with visualisation exercises:

<img src="https://imgur.com/qYIvTx1.jpg" width="500" height="211" />

### Conclusion

We hope that our converted assignments will help scale the course and deliver something very close to the experience we had last fall. 
Both of us learned a lot through this project and we are very eager to hear what next years students have to say about the course! We would like to thank Dr. Çetinkaya-Rundel for the giving us the opportunity to be part of IDS and her neverending patience for us! 

> If you're interested, one of our tutorials was slightly adapted for use the Teach R Online workshop series. You can see it [here](https://gallery.shinyapps.io/lego-sales/).

Reviewer: @lee-suddaby you can also add a sentence here that we're placing them in the dsbox package and that they can be ran directly from the tutorials pane of RStudio, and maybe we can even add a screenshot with the tutorials showing up.

---
title: Data Science Tutorials with learnr and gradethis
author: Lee Suddaby, Zeno Kujawa
date: '2020-07-24'
slug: data-science-tutorials-with-learnr-and-gradethis
categories: 
  - teaching
  - shiny
  - rstats
tags: 
  - learnr
  - gradethis
  - tidyverse
keywords:
  - tech
---

> *This post was contributed by [Lee Suddaby](https://github.com/lee-suddaby) and [Zeno Kujawa](https://github.com/ZenoMK), second year students at the University of Edinburgh majoring in Mathematics and Data Science, respectively.*

Author: @ZenoMK If you want to change the introduction for yourself  (or any other part of the post of course) just do so. Feel free to change the link to your profile if there's something else that's more appropriate - I wasn't sure when I wrote it.

![Learnr tutorial code block showing visualisation exercise example](/post/2020-07-24-data-science-tutorials-with-learnr-and-gradethis_files/qYIvTx1.png)

### Conclusion

We hope that our converted assignments will help scale the course and deliver something very close to the experience we had last fall. 
Both of us learned a lot through this project and we are very eager to hear what next years students have to say about the course! We would like to thank Dr. Çetinkaya-Rundel for the giving us the opportunity to be part of IDS and her neverending patience for us! 

For students to access the tutorials, we're placing them in the [**dsbox**](https://www.github.com/rstudio-education/dsbox) package. They can then be run directly from the *Tutorials* pane in RStudio, like so:

> ![Tutorials pane in RStudio](/post/2020-07-24-data-science-tutorials-with-learnr-and-gradethis_files/tutorials-pane.PNG)

Author: @mine-cetinkaya-rundel I've got a screenshot for the tutorials pane. Unfortunately, neither RStudio cloud nor my local installation seems to want to actually run one of the tutorials themselves, (it would appear my talents are limited in certain areas), so if we want more than just a screenshot of the tutorials pane you might need to be the one to take it. sorry... (I am trying to get it working, promise!)
